TheAmazingAudioEngine/AEAudioUnitFileStreamer.m

```diff
@@ -0,0 +1,395 @@
+        // we don't want to start playing immediately.
+        self.channelIsPlaying = NO;
+        
+        _url = url;
+    }
+    
+    return(self);
+}
+
+- (BOOL) setupURL: (NSURL*) url error:(NSError**) error{
+    // Load the URL
+    OSStatus result;
+    ExtAudioFileRef audioFileRef;
+    checkResult(result = ExtAudioFileOpenURL((CFURLRef)CFBridgingRetain(url), &audioFileRef), "AudioFileOpenURL");
+    if(result == noErr)
+    {
+        // Get the audio file itself
+        UInt32 size = sizeof(_audioUnitFile);
+        checkResultForError(ExtAudioFileGetProperty(audioFileRef, kExtAudioFileProperty_AudioFile, &size, &_audioUnitFile), "ExtAudioFileGetProperty(kExtAudioFileProperty_AudioFile)", "Failed getting audio file from extAudioFileRef", NO, result, error)
+        
+        // Set the file to play
+        AudioUnit au = AEAudioUnitChannelGetAudioUnit(self);
+        checkResultForError(AudioUnitSetProperty(au, kAudioUnitProperty_ScheduledFileIDs, kAudioUnitScope_Global, 0, &_audioUnitFile, sizeof(_audioUnitFile)), "AudioUnitSetProperty(kAudioUnitProperty_ScheduledFileIDs)", "Failed setting audio unit file IDs", NO, result, error)
+        
+        // Get audio description
+        size = sizeof(_audioDescription);
+        checkResultForError(AudioFileGetProperty(_audioUnitFile, kAudioFilePropertyDataFormat, &size, &_audioDescription), "AudioFileGetProperty(kAudioFilePropertyDataFormat)", "Failed getting audio file data format", NO, result, error)
+        
+        // Get file length
+        UInt64 fileLengthInFrames = 0;
+        size = sizeof(fileLengthInFrames);
+        checkResultForError(ExtAudioFileGetProperty(audioFileRef, kExtAudioFileProperty_FileLengthFrames, &size, &fileLengthInFrames), "ExtAudioFileGetProperty(kExtAudioFileProperty_AudioFile)", "Failed getting file length", NO, result, error)
+        
+        _lengthInFrames = (UInt32)fileLengthInFrames;
+        _url = url;
+        
+        return(YES);
+    }
+    else {
+        *error = [NSError errorWithDomain:NSOSStatusErrorDomain code:result userInfo:[NSDictionary dictionaryWithObject:@"URL failed to open" forKey:NSLocalizedDescriptionKey]];
+        return(NO);
+    }
+}
+
+- (void)dealloc {
+    
+    if ( _audioUnitFile ) {
+        AudioFileClose(_audioUnitFile);
+    }
+    
+    _completionBlock = nil;
+    _audioController = nil;
+}
+
+// MARK: Audio functions
+
+- (void) play
+{
+    self.channelIsPlaying = true;
+}
+- (void) pause
+{
+    self.channelIsPlaying = false;
+}
+- (void) stop
+{
+    self.channelIsPlaying = NO;
+    _playhead = 0;
+    NSError* error;
+    [self setupPlayRegion:&error];
+    if(error)
+        NSLog(@"Error on playback stopped callback: %@",error.description);
+}
+
+- (void) mute
+{
+    self.channelIsMuted = true;
+}
+
+- (void) unmute
+{
+    self.channelIsMuted = false;
+}
+
+// MARK: Getters and Setters
+-(NSTimeInterval)duration {
+    return (double)_lengthInFrames / (double)_audioDescription.mSampleRate;
+}
```

Comment:
Seems that the easiest fix is to move the _lengthInFrames calculation out of setupURL:error: and into the initializer.

I've never run into this issue because I always add a channel immediately after creating it. Then set channelIsPlaying (or call the play function on this class) when ready to play. Is there any reason not to add a channel immediately? Should there be a TAAE best practice regarding this?